pkg/controller/volume/selinuxwarning/selinux_warning_controller.go

```diff
@@ -0,0 +1,507 @@
+	errorutils "k8s.io/apimachinery/pkg/util/errors"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	storageinformersv1 "k8s.io/client-go/informers/storage/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	storagelisters "k8s.io/client-go/listers/storage/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	csitrans "k8s.io/csi-translation-lib"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/controller/volume/attachdetach/util"
+	"k8s.io/kubernetes/pkg/controller/volume/common"
+	volumecache "k8s.io/kubernetes/pkg/controller/volume/selinuxwarning/cache"
+	"k8s.io/kubernetes/pkg/volume"
+	"k8s.io/kubernetes/pkg/volume/csi"
+	"k8s.io/kubernetes/pkg/volume/csimigration"
+	volumeutil "k8s.io/kubernetes/pkg/volume/util"
+	"k8s.io/utils/ptr"
+)
+
+// SELinuxWarning controller is a controller that emits a warning event and metrics when
+// two pods *might* use the same volume with different SELinux labels.
+// It is optional. It does nothing on a cluster that has SELinux disabled.
+// It does not modify any API objects except for the warning events.
+//
+// The controller watches *all* Pods and reports issues on all their volumes,
+// regardless if the Pod actually run on are Pending forever or if they have
+// correct node anti-affinity and never run on the same node.
+type Controller struct {
+	kubeClient       clientset.Interface
+	podLister        corelisters.PodLister
+	podIndexer       cache.Indexer
+	podsSynced       cache.InformerSynced
+	pvLister         corelisters.PersistentVolumeLister
+	pvsSynced        cache.InformerSynced
+	pvcLister        corelisters.PersistentVolumeClaimLister
+	pvcsSynced       cache.InformerSynced
+	csiDriverLister  storagelisters.CSIDriverLister
+	csiDriversSynced cache.InformerSynced
+
+	vpm               *volume.VolumePluginMgr
+	cmpm              csimigration.PluginManager
+	csiTranslator     csimigration.InTreeToCSITranslator
+	seLinuxTranslator volumeutil.SELinuxLabelTranslator
+	eventRecorder     record.EventRecorder
+
+	recorder record.EventRecorder
+	queue    workqueue.TypedRateLimitingInterface[cache.ObjectName]
+
+	labelCache volumecache.VolumeCache
+}
+
+func NewController(
+	ctx context.Context,
+	kubeClient clientset.Interface,
+	podInformer coreinformers.PodInformer,
+	pvcInformer coreinformers.PersistentVolumeClaimInformer,
+	pvInformer coreinformers.PersistentVolumeInformer,
+	csiDriverInformer storageinformersv1.CSIDriverInformer,
+	plugins []volume.VolumePlugin,
+	prober volume.DynamicPluginProber,
+) (*Controller, error) {
+
+	eventBroadcaster := record.NewBroadcaster(record.WithContext(ctx))
+	eventBroadcaster.StartLogging(klog.Infof)
```

Comment:
I poked around, it starts watching an event broadcaster and whatnot. I moved it to Run, just in case someone instantiates the controller but does not run it, and changed to `StartStructuredLogging`. And changed the log level of events to `V(3)`, as other controllers do.